File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/reactiveobject/UaReadEventHistoryTransaction.java
package org.opcfoundation.webserver.service.transactions.reactiveobject;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject;
import org.eclipse.milo.opcua.stack.core.types.structured.HistoryEvent;
import org.eclipse.milo.opcua.stack.core.types.structured.HistoryEventFieldList;
import org.opcfoundation.webapi.mapper.extensionobjects.ExtensionObjectEncoder;
import org.opcfoundation.webapi.service.types.HistoryReadContext;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadHistoryEventRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadHistoryEventResponse;
import org.opcfoundation.webserver.service.transactions.base.UaHistoryReadTransaction;
import org.opcfoundation.webserver.types.common.UaHistoryReadContinuationPoint;
import org.opcfoundation.webserver.types.common.UaInstanceIdentifier;
import org.opcfoundation.webserver.types.common.UaObjectId;
import java.util.concurrent.CompletableFuture;

public class UaReadEventHistoryTransaction extends UaHistoryReadTransaction {
    private final NodeManagerReactiveObject nodeManager;
    private final UaInstanceIdentifier objectIdentifier;
    private final int offset;

    public UaReadEventHistoryTransaction(
            HistoryReadContext historyReadContext,
            int index,
            UaInstanceIdentifier objectIdentifier,
            NodeManagerReactiveObject nodeManager)
    {
        super(historyReadContext, index);
        this.objectIdentifier = objectIdentifier;
        this.nodeManager = nodeManager;

        if (getItem().getContinuationPoint().isNull())
        {
            offset = 0;
        } else {
            UaHistoryReadContinuationPoint cp = UaHistoryReadContinuationPoint.fromByteString(getItem().getContinuationPoint());
            if (null == cp)
            {
                offset = -1;
            } else {
                offset = cp.getOffset();
            }
        }
    }

    public CompletableFuture<Void> execute()
    {
        try
        {
            UaReactiveObjectType objectType = nodeManager.findObjectType(objectIdentifier.getObjectId());
            UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(objectIdentifier.getObjectId());
            if (null == objectType) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);
            if (0 > offset) throw new UaRuntimeException(StatusCodes.Bad_ContinuationPointInvalid);

            ReadHistoryEventRequest request = new ReadHistoryEventRequest(
                    new UaObjectId(objectIdentifier.getObjectId().getId(), instanceDeclaration),
                    details,
                    offset);

            return objectType.onReadHistoryEvent(request).
                    thenAccept(this::setResults).
                    exceptionally(ex -> buildErrorResponse(ex.getCause()));
        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private void setResults(ReadHistoryEventResponse response) {
        HistoryEvent historyEvent = new HistoryEvent(response.getEvents().toArray(new HistoryEventFieldList[0]));
        historyData = ExtensionObject.encode(ExtensionObjectEncoder.Encoder.getEncodingContext(),historyEvent);

        if (response.getContainsMoreData() && !response.getEvents().isEmpty())
        {
            UaHistoryReadContinuationPoint cp = new UaHistoryReadContinuationPoint();
            cp.setOffset(offset + response.getEvents().size());
            continuationPoint = cp.toByteString();
        }
    }
}

File: opcua-webservice-node/source/src/server/service/transactions/reactiveobject/UaReadEventHistoryTransaction.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaError, UaHistoryEvent } from "opcua-webapi-ts";
import { HistoryReadContext } from "../../..";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaHistoryReadContinuationPoint, UaInstanceIdentifier, UaObjectId } from "../../../types";
import { ReadHistoryEventRequest, ReadHistoryEventResponse } from "../../message";
import { UaHistoryReadTransaction } from "../base/UaHistoryReadTransaction";

export class UaReadEventHistoryTransaction extends UaHistoryReadTransaction {
    private readonly nodeManager: NodeManagerReactiveObject;
    private readonly objectIdentifier: UaInstanceIdentifier;
    private readonly offset: number;

    constructor(
        historyReadContext: HistoryReadContext,
        index: number,
        objectIdentifier: UaInstanceIdentifier,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(historyReadContext, index);
        this.objectIdentifier = objectIdentifier;
        this.nodeManager = nodeManager;

        const continuationPoint = this.getItem().continuationPoint;
        if (!continuationPoint) {
            this.offset = 0;
            return;
        }

        const parsedContinuationPoint = UaHistoryReadContinuationPoint.fromByteString(continuationPoint);
        this.offset = parsedContinuationPoint?.offset ?? -1;
    }

    override async execute(): Promise<void> {
        try {
            const objectType = this.nodeManager.findObjectType(this.objectIdentifier.objectId);
            const instanceDeclaration = this.nodeManager.findInstanceDeclaration(this.objectIdentifier.objectId);

            if (objectType === null) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            }

            if (this.offset < 0) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadContinuationPointInvalid));
            }

            const request = new ReadHistoryEventRequest(
                new UaObjectId(this.objectIdentifier.objectId.id, instanceDeclaration),
                this.details,
                this.offset,
            );

            const response = await objectType.onReadHistoryEvent(request);
            this.setResults(response);
        } catch (error) {
            this._historyData = undefined;
            this._continuationPoint = undefined;
            this._statusCode = makeUaStatusCode(StatusCodes.BadUnexpectedError);

            if (error instanceof UaError) {
                this._statusCode = error.statusCode;
            }
        }
    }

    private setResults(response: ReadHistoryEventResponse): void {
        this._historyData = new UaHistoryEvent(response.events).toExtensionObject();
        this._continuationPoint = undefined;

        if (response.containsMoreData && response.events.length > 0) {
            this._continuationPoint = new UaHistoryReadContinuationPoint(
                this.offset + response.events.length).toByteString();
        }
    }
}